Formatting error in complex agrument types

## Configuration (required) ##

```
version = 3.8.3
runner.dialect = scala213source3
preset = default

newlines.source = keep
align.openParenCallSite = true
```

## Command-line parameters (required) ##

When I run scalafmt via CLI like this: `scalafmt A.scala`

## Steps

Given code like this:

```scala
package foo

object A {
  def f(a: (
             Seq[Int],
               Seq[Int]) =
           (Seq(), Seq())): Unit = ???
}
```

## Problem

Scalafmt formats code like this:

```shell
> scalafmt <path>/A.scala
org.scalafmt.cli.FailedToFormat: <path>/A.scala
Caused by: org.scalafmt.Error$PreciseIncomplete: <path>/A.scala:4: error: Unable to format file due to bug in scalafmt
  def f(a: (
           ^
	at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$6(Scalafmt.scala:123)
	at scala.util.Success.flatMap(Try.scala:266)
	at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$3(Scalafmt.scala:118)
	at scala.meta.parsers.Parsed.fold(Errors.scala:11)
	at scala.meta.parsers.Parsed.fold$(Errors.scala:10)
	at scala.meta.parsers.Parsed$Success.fold(Errors.scala:25)
	at org.scalafmt.Scalafmt$.doFormatOne(Scalafmt.scala:114)
	at org.scalafmt.Scalafmt$.doFormat(Scalafmt.scala:90)
	at org.scalafmt.Scalafmt$.$anonfun$formatCode$3(Scalafmt.scala:78)
	at scala.util.Success.fold(Try.scala:289)
	at org.scalafmt.Scalafmt$.formatCode(Scalafmt.scala:78)
	at org.scalafmt.Scalafmt$.format(Scalafmt.scala:57)
	at org.scalafmt.Scalafmt.format(Scalafmt.scala)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.scalafmt.dynamic.ScalafmtReflect.$anonfun$tryFormat$1(ScalafmtReflect.scala:121)
	at scala.util.Try$.apply(Try.scala:217)
	at org.scalafmt.dynamic.ScalafmtReflect.tryFormat(ScalafmtReflect.scala:117)
	at org.scalafmt.dynamic.ScalafmtReflectConfig.tryFormat(ScalafmtReflectConfig.scala:95)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.$anonfun$tryForceFormat$1(ScalafmtDynamicSession.scala:49)
	at scala.util.Success.flatMap(Try.scala:266)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.tryForceFormat(ScalafmtDynamicSession.scala:49)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.tryFormat(ScalafmtDynamicSession.scala:36)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.format(ScalafmtDynamicSession.scala:18)
	at org.scalafmt.cli.ScalafmtDynamicRunner$.handleFile(ScalafmtDynamicRunner.scala:74)
	at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$7(ScalafmtDynamicRunner.scala:48)
	at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$7$adapted(ScalafmtDynamicRunner.scala:46)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:619)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:617)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1303)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:938)
	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:52)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:97)
	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:55)
	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:49)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:935)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute(Tasks.scala:152)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute$(Tasks.scala:148)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$AWSFJTWrappedTask.compute(Tasks.scala:304)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```


## Expectation

## Workaround

If either of these lines are removed, it will not reproduce

```
newlines.source = keep
align.openParenCallSite = true
```

## Notes

This code also raises an error.

```scala
package foo

object A {
//  def f(a: (
//             Seq[Int],
//               Seq[Int]) =
//           (Seq(), Seq())): Unit = ???

  private def g: Either[
    (
      Int, Option[Int]),
    (Int, Option[Int])
  ] = ???
}
```

```shell
> scalafmt <path>/A.scala
org.scalafmt.cli.FailedToFormat: <path>/A.scala
Caused by: org.scalafmt.Error$PreciseIncomplete: <path>/A.scala:10: error: Unable to format file due to bug in scalafmt
    (
    ^
	at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$6(Scalafmt.scala:123)
	at scala.util.Success.flatMap(Try.scala:266)
	at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$3(Scalafmt.scala:118)
	at scala.meta.parsers.Parsed.fold(Errors.scala:11)
	at scala.meta.parsers.Parsed.fold$(Errors.scala:10)
	at scala.meta.parsers.Parsed$Success.fold(Errors.scala:25)
	at org.scalafmt.Scalafmt$.doFormatOne(Scalafmt.scala:114)
	at org.scalafmt.Scalafmt$.doFormat(Scalafmt.scala:90)
	at org.scalafmt.Scalafmt$.$anonfun$formatCode$3(Scalafmt.scala:78)
	at scala.util.Success.fold(Try.scala:289)
	at org.scalafmt.Scalafmt$.formatCode(Scalafmt.scala:78)
	at org.scalafmt.Scalafmt$.format(Scalafmt.scala:57)
	at org.scalafmt.Scalafmt.format(Scalafmt.scala)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.scalafmt.dynamic.ScalafmtReflect.$anonfun$tryFormat$1(ScalafmtReflect.scala:121)
	at scala.util.Try$.apply(Try.scala:217)
	at org.scalafmt.dynamic.ScalafmtReflect.tryFormat(ScalafmtReflect.scala:117)
	at org.scalafmt.dynamic.ScalafmtReflectConfig.tryFormat(ScalafmtReflectConfig.scala:95)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.$anonfun$tryForceFormat$1(ScalafmtDynamicSession.scala:49)
	at scala.util.Success.flatMap(Try.scala:266)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.tryForceFormat(ScalafmtDynamicSession.scala:49)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.tryFormat(ScalafmtDynamicSession.scala:36)
	at org.scalafmt.dynamic.ScalafmtDynamicSession.format(ScalafmtDynamicSession.scala:18)
	at org.scalafmt.cli.ScalafmtDynamicRunner$.handleFile(ScalafmtDynamicRunner.scala:74)
	at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$7(ScalafmtDynamicRunner.scala:48)
	at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$7$adapted(ScalafmtDynamicRunner.scala:46)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:619)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:617)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1303)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:938)
	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:52)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:97)
	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:55)
	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:49)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:935)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute(Tasks.scala:152)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute$(Tasks.scala:148)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$AWSFJTWrappedTask.compute(Tasks.scala:304)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

Please note that this/these problem was not caused by intentionally giving pathological code to cause errors, but was accidental due to the corrupted formatting that existed in our project.

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
@@ -2567,6 +2567,14 @@ class Router(formatOps: FormatOps) {
     */
   def getSplits(ft: FormatToken): Seq[Split] = {
     val splits = getSplitsImpl(ft).filter(!_.isIgnored)
+    require(
+      splits.nonEmpty,
+      s"""|
+          |FT: ${log2(ft)}
+          |LO: ${log(ft.leftOwner)}
+          |RO: ${log(ft.rightOwner)}
+          |""".stripMargin,
+    )
     def splitsAsNewlines(splits: Seq[Split]): Seq[Split] = {
       val filtered = Decision.onlyNewlineSplits(splits)
       if (filtered.nonEmpty) filtered else splits.map(_.withMod(Newline))

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
@@ -940,8 +940,13 @@ class Router(formatOps: FormatOps) {
           (if (onlyConfigStyle) opensConfigStyleImplicitParamList(ft)
            else hasImplicitParamList(rightOwner))
 
+        val rightIsComment = right.is[T.Comment]
+        val align = !rightIsComment && alignOpenDelim &&
+          (!handleImplicit || style.newlines.forceAfterImplicitParamListModifier)
+        val alignTuple = align && tupleSite && !onlyConfigStyle
+
         val noSplitForNL = !onlyConfigStyle && right.is[T.LeftBrace]
-        val skipNoSplit = !noSplitForNL &&
+        val skipNoSplit = !noSplitForNL && !alignTuple &&
           (style.newlines.keepBreak(newlines) || {
             if (!handleImplicit) onlyConfigStyle
             else style.newlines.forceBeforeImplicitParamListModifier
@@ -949,12 +954,6 @@ class Router(formatOps: FormatOps) {
         val noSplitMod =
           if (skipNoSplit) null
           else getNoSplitAfterOpening(ft, commentNL = null, spaceOk = !isBracket)
-
-        val rightIsComment = right.is[T.Comment]
-
-        val align = !rightIsComment && alignOpenDelim &&
-          (!handleImplicit || style.newlines.forceAfterImplicitParamListModifier)
-        val alignTuple = align && tupleSite && !onlyConfigStyle
 
         val keepConfigStyleSplit = !sourceIgnored && configStyleFlag &&
           hasBreak()
@@ -1025,8 +1024,8 @@ class Router(formatOps: FormatOps) {
             )
           }
 
-        val keepNoNL = style.newlines.keepBreak(noBreak())
-        val preferNoSplit = keepNoNL && singleArgument
+        val preferNoSplit = !skipNoSplit && singleArgument &&
+          style.newlines.keepBreak(noBreak())
         val oneArgOneLine = newlinePolicy &
           (leftOwner match {
             case t @ (_: Member.SyntaxValuesClause | _: Member.Tuple) =>

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
@@ -1096,8 +1096,9 @@ class Router(formatOps: FormatOps) {
 
         val splitsNL =
           if (
-            alignTuple ||
-            !(onlyConfigStyle || multipleArgs || splitsForAssign.isEmpty)
+            splitsNoNL.nonEmpty &&
+            (alignTuple ||
+              !(onlyConfigStyle || multipleArgs || splitsForAssign.isEmpty))
           ) Seq.empty
           else {
             val cost =

--- a/scalafmt-tests/shared/src/test/resources/newlines/source_classic.stat
+++ b/scalafmt-tests/shared/src/test/resources/newlines/source_classic.stat
@@ -10405,3 +10405,20 @@ object a {
       .map(filterNot(getCanonicalPath.contains("SSLOptions")))
   }
 }
+<<< #4318
+runner.dialect = scala213source3
+align.openParenCallSite = true
+===
+object A {
+  def f(a: (
+             Seq[Int],
+               Seq[Int]) =
+           (Seq(), Seq())): Unit = ???
+}
+>>>
+object A {
+  def f(
+      a: (Seq[Int], Seq[Int]) =
+        (Seq(), Seq())
+  ): Unit = ???
+}

--- a/scalafmt-tests/shared/src/test/resources/newlines/source_fold.stat
+++ b/scalafmt-tests/shared/src/test/resources/newlines/source_fold.stat
@@ -9712,3 +9712,21 @@ object a {
       .map(filterNot(getCanonicalPath.contains("SSLOptions")))
   }
 }
+<<< #4318
+runner.dialect = scala213source3
+align.openParenCallSite = true
+newlines.configStyleCallSite.prefer = false
+===
+object A {
+  def f(a: (
+             Seq[Int],
+               Seq[Int]) =
+           (Seq(), Seq())): Unit = ???
+}
+>>>
+object A {
+  def f(
+      a: (Seq[Int], Seq[Int]) =
+        (Seq(), Seq())
+  ): Unit = ???
+}

--- a/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
+++ b/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
@@ -10157,7 +10157,7 @@ object a {
       .map(filterNot(getCanonicalPath.contains("SSLOptions")))
   }
 }
-<<< SKIP #4318
+<<< #4318
 runner.dialect = scala213source3
 align.openParenCallSite = true
 ===
@@ -10168,16 +10168,10 @@ object A {
            (Seq(), Seq())): Unit = ???
 }
 >>>
-org.scalafmt.util.FormatException: java.lang.IllegalArgumentException: requirement failed: 
-FT: [9] (∙Seq: LeftParen [22..23) [LF] Ident(Seq) [37..40)
-LO: TYPE: Type$Tuple$TypeTupleImpl
-SOURCE: (
-             Seq[Int],
-               Seq[Int])
-STRUCTURE: Type.Tuple(List(Type.Apply(Type.Name("Seq"), Type.ArgClause(List(Type.Name("Int")))), Type.Apply(Type.Name("Seq"), Type.ArgClause(List(Type.Name("Int"))))))
-TOKENS: (,¶,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,Seq,[,Int,],,,¶,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,Seq,[,Int,],)
-
-RO: TYPE: Type$Name$TypeNameImpl
-SOURCE: Seq
-STRUCTURE: Type.Name("Seq")
-TOKENS: Seq
+object A {
+  def f(a: (
+      Seq[Int],
+      Seq[Int]
+  ) =
+    (Seq(), Seq())): Unit = ???
+}

--- a/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
+++ b/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
@@ -10168,12 +10168,16 @@ object A {
            (Seq(), Seq())): Unit = ???
 }
 >>>
-BestFirstSearch:310 Failed to format
-UNABLE TO FORMAT,
-tok #9/36: [9] (∙Seq: LeftParen [22..23) [LF] Ident(Seq) [37..40)
-policies:
-  PNL+2:[Router:1068]<100d
-  NB:[Router:243]@114d
-splits (before policy):
-  
-splits (after policy):
+org.scalafmt.util.FormatException: java.lang.IllegalArgumentException: requirement failed: 
+FT: [9] (∙Seq: LeftParen [22..23) [LF] Ident(Seq) [37..40)
+LO: TYPE: Type$Tuple$TypeTupleImpl
+SOURCE: (
+             Seq[Int],
+               Seq[Int])
+STRUCTURE: Type.Tuple(List(Type.Apply(Type.Name("Seq"), Type.ArgClause(List(Type.Name("Int")))), Type.Apply(Type.Name("Seq"), Type.ArgClause(List(Type.Name("Int"))))))
+TOKENS: (,¶,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,Seq,[,Int,],,,¶,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,∙,Seq,[,Int,],)
+
+RO: TYPE: Type$Name$TypeNameImpl
+SOURCE: Seq
+STRUCTURE: Type.Name("Seq")
+TOKENS: Seq

--- a/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
+++ b/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
@@ -10157,3 +10157,23 @@ object a {
       .map(filterNot(getCanonicalPath.contains("SSLOptions")))
   }
 }
+<<< SKIP #4318
+runner.dialect = scala213source3
+align.openParenCallSite = true
+===
+object A {
+  def f(a: (
+             Seq[Int],
+               Seq[Int]) =
+           (Seq(), Seq())): Unit = ???
+}
+>>>
+BestFirstSearch:310 Failed to format
+UNABLE TO FORMAT,
+tok #9/36: [9] (∙Seq: LeftParen [22..23) [LF] Ident(Seq) [37..40)
+policies:
+  PNL+2:[Router:1068]<100d
+  NB:[Router:243]@114d
+splits (before policy):
+  
+splits (after policy):

--- a/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
+++ b/scalafmt-tests/shared/src/test/resources/newlines/source_keep.stat
@@ -10169,9 +10169,6 @@ object A {
 }
 >>>
 object A {
-  def f(a: (
-      Seq[Int],
-      Seq[Int]
-  ) =
+  def f(a: (Seq[Int], Seq[Int]) =
     (Seq(), Seq())): Unit = ???
 }

--- a/scalafmt-tests/shared/src/test/resources/newlines/source_unfold.stat
+++ b/scalafmt-tests/shared/src/test/resources/newlines/source_unfold.stat
@@ -10512,3 +10512,21 @@ object a {
       .map(filterNot(getCanonicalPath.contains("SSLOptions")))
   }
 }
+<<< #4318
+runner.dialect = scala213source3
+align.openParenCallSite = true
+newlines.configStyleCallSite.prefer = false
+===
+object A {
+  def f(a: (
+             Seq[Int],
+               Seq[Int]) =
+           (Seq(), Seq())): Unit = ???
+}
+>>>
+object A {
+  def f(
+      a: (Seq[Int], Seq[Int]) =
+        (Seq(), Seq())
+  ): Unit = ???
+}

--- a/scalafmt-tests/shared/src/test/scala/org/scalafmt/FormatTests.scala
+++ b/scalafmt-tests/shared/src/test/scala/org/scalafmt/FormatTests.scala
@@ -137,7 +137,7 @@ class FormatTests extends FunSuite with CanRunTests with FormatAssertions {
   override def afterAll(): Unit = {
     logger.debug(s"Total explored: ${Debug.explored}")
     if (!onlyUnit && !onlyManual)
-      assertEquals(Debug.explored, 1784910, "total explored")
+      assertEquals(Debug.explored, 1784986, "total explored")
     val results = debugResults.result()
     // TODO(olafur) don't block printing out test results.
     // I don't want to deal with scalaz's Tasks :'(

--- a/scalafmt-tests/shared/src/test/scala/org/scalafmt/FormatTests.scala
+++ b/scalafmt-tests/shared/src/test/scala/org/scalafmt/FormatTests.scala
@@ -137,7 +137,7 @@ class FormatTests extends FunSuite with CanRunTests with FormatAssertions {
   override def afterAll(): Unit = {
     logger.debug(s"Total explored: ${Debug.explored}")
     if (!onlyUnit && !onlyManual)
-      assertEquals(Debug.explored, 1784986, "total explored")
+      assertEquals(Debug.explored, 1784976, "total explored")
     val results = debugResults.result()
     // TODO(olafur) don't block printing out test results.
     // I don't want to deal with scalaz's Tasks :'(

--- a/scalafmt-tests/shared/src/test/scala/org/scalafmt/FormatTests.scala
+++ b/scalafmt-tests/shared/src/test/scala/org/scalafmt/FormatTests.scala
@@ -137,7 +137,7 @@ class FormatTests extends FunSuite with CanRunTests with FormatAssertions {
   override def afterAll(): Unit = {
     logger.debug(s"Total explored: ${Debug.explored}")
     if (!onlyUnit && !onlyManual)
-      assertEquals(Debug.explored, 1784403, "total explored")
+      assertEquals(Debug.explored, 1784910, "total explored")
     val results = debugResults.result()
     // TODO(olafur) don't block printing out test results.
     // I don't want to deal with scalaz's Tasks :'(

